fdbserver/workloads/TenantCapacityLimits.actor.cpp

```diff
@@ -0,0 +1,194 @@
+#include "fdbrpc/simulator.h"
+#include "fdbserver/workloads/MetaclusterConsistency.actor.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "fdbserver/Knobs.h"
+#include "flow/BooleanParam.h"
+#include "flow/Error.h"
+#include "flow/IRandom.h"
+#include "flow/ThreadHelper.actor.h"
+#include "flow/Trace.h"
+#include "flow/flow.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+struct TenantCapacityLimits : TestWorkload {
+	static constexpr auto NAME = "TenantCapacityLimits";
+
+	Reference<IDatabase> managementDb;
+	Database dataDb;
+
+	int64_t tenantIdPrefix;
+	bool useMetacluster = false;
+	const Key specialKeysTenantMapPrefix = SpecialKeySpace::getModuleRange(SpecialKeySpace::MODULE::MANAGEMENT)
+	                                           .begin.withSuffix(TenantRangeImpl::submoduleRange.begin)
+	                                           .withSuffix(TenantRangeImpl::mapSubRange.begin);
+
+	TenantCapacityLimits(WorkloadContext const& wcx) : TestWorkload(wcx) {
+		tenantIdPrefix = getOption(options,
+		                           "tenantIdPrefix"_sr,
+		                           deterministicRandom()->randomInt(TenantAPI::TENANT_ID_PREFIX_MIN_VALUE,
+		                                                            TenantAPI::TENANT_ID_PREFIX_MAX_VALUE + 1));
+		if (clientId == 0) {
+			useMetacluster = deterministicRandom()->coinflip();
+		}
+	}
+
+	void disableFailureInjectionWorkloads(std::set<std::string>& out) const override { out.insert("Attrition"); }
+
+	Future<Void> setup(Database const& cx) override {
+		if (clientId == 0) {
+			if (g_network->isSimulated() && BUGGIFY) {
+				IKnobCollection::getMutableGlobalKnobCollection().setKnob(
+				    "max_tenants_per_cluster", KnobValueRef::create(int{ deterministicRandom()->randomInt(20, 100) }));
+			}
+			return _setup(cx, this);
+		} else {
+			return Void();
+		}
+	}
+	ACTOR static Future<Void> _setup(Database cx, TenantCapacityLimits* self) {
+		if (self->useMetacluster) {
+			Reference<IDatabase> threadSafeHandle =
+			    wait(unsafeThreadFutureToFuture(ThreadSafeDatabase::createFromExistingDatabase(cx)));
+
+			MultiVersionApi::api->selectApiVersion(cx->apiVersion.version());
+			self->managementDb = MultiVersionDatabase::debugCreateFromExistingDatabase(threadSafeHandle);
+
+			wait(success(
+			    MetaclusterAPI::createMetacluster(cx.getReference(), "management_cluster"_sr, self->tenantIdPrefix)));
+
+			DataClusterEntry entry;
+			entry.capacity.numTenantGroups = 1e9;
+			wait(MetaclusterAPI::registerCluster(
+			    self->managementDb, "test_data_cluster"_sr, g_simulator->extraDatabases[0], entry));
+
+			ASSERT(g_simulator->extraDatabases.size() == 1);
+			self->dataDb = Database::createSimulatedExtraDatabase(g_simulator->extraDatabases[0], cx->defaultTenant);
+			// wait for tenant mode change on dataDB
+			wait(success(self->waitDataDbTenantModeChange()));
+		} else {
+			self->dataDb = cx;
+		}
+
+		return Void();
```

Comment:
As a random aside, maybe we should make a class that we could inherit from to make it easier to write metacluster tests.

Reply:
Sure but i would need to assess all other workload uses, I dont know if it makes sense to do it in this pr, if time permitting I can do it afterwards

Reply:
Yeah I wouldn't suggest doing that in this PR, it was more an observation